Restore inline site preview

davemart-in opened the pull request:
## Background

In the latest nav drawer update we made the decision to remove the inline site preview screen. My rationale for removing it was based on a number of usability concerns that have cropped up in the past with that screen:

- People wondering if their site visitors will see all of this navigation when visiting their site
- Having multiple top nav bars stacked on top of each other
- Needing to click “View Site” in the left nav, then click the “View Site” primary button in the top right corner to actually get to your site.

After deploying this change to 100% of users, @carinapilar brought up an instance of a customer that was upset by this functionality being taken away.

@shaunandrews then weighed in with some points that I had not considered when we removed this screen:

1. With the front-end menu, things are really complicated — there's no obvious way back.
2. Viewing your site at different breakpoints is gone.
3. Viewing your SEO preview is gone. This is a big one — as its something people explicitly upgrade for, and that we advertising in our messaging.
4. Personal attachment to the site preview component. We had great ambitions for continued improvement, including tighter integration with sharing tools.
5. We lost the ability to show our `inlineHelp` component and Guided Tours.
6. Plus, there are some technical advantages to loading your site in the Calypso context, including some performance improvements that make loading your site in the preview faster than the front-end.

Weighing these additional insights, it became clear that the pros of removing it probably don't outweigh the cons. I offered to take a stab at creating a PR to add it back.

## Changes

Visibly, not much has changed. 

Prior to this change, when you clicked the site card, you were taken directly to the front-end of your site. Now you're taken to the inline view site preview.

## Notes

1) I initially tried to commit this as a `<a>` tag, but linting yelled at me:

```
/Users/davem/Projects/Automattic/wp-calypso/client/blocks/site/index.jsx
125:5  error  Anchor used as a button. Anchors are primarily expected to navigate. Use the button element instead. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md  jsx-a11y/anchor-is-valid
```

So I changed it to a button.

2) I had to adjust some of the CSS as a result of switching this from a link to a button.

3) I mostly just moved over the code that @apeatling removed in https://github.com/Automattic/wp-calypso/pull/31616/files

4) I was unsure whether we really needed `( isSingleSite && getPrimarySiteId( state ) )` as a fallback for `siteId`. If we got rid of it, we could eliminate `isSingleSite`, and `getPrimarySiteId`, and as a result `currentUser` and `getCurrentUser`. I'll let someone who knows better weigh in here.

## Testing

Should be pretty straightforward. Let me know if you have any questions or concerns. Thanks!

Commit:
Initial stab at restoring the inline site preview

Commit:
Added pointer cursor style

Commit:
Added focus box-shadow

jsnajdr commented:
> I was unsure whether we really needed ( isSingleSite && getPrimarySiteId( state ) ) as a fallback for siteId.

You copied some code from the `<Sidebar>` component, didn't you? For `<Site>` component, this is not needed at all (and actually breaks a lot of things).

`<Sidebar>` is rendered without any props that would tell it which site is selected. So it needs to figure it out on its own by reading from Redux state. The logic is: get the currently selected site, and if none is selected, get the user's primary site.

On the other hand, the `<Site>` component receives `siteId` or `site` prop that tells it which site to use. It doesn't need to figure it out. Your code actually ignores and overwrites the site passed in props and renders a completely different one!

What site needs to figure out are some properties of the requested site. Is it previewable? What is it's site slug? It uses Redux selectors and the site ID to find that. With calls like `isSitePreviewable( state, siteId )`.

There are two ways to use the `<Site>` component:
- by passing only the `siteId` (number) as prop and it will read the site object with `getSite`.
- by passing the full `site` object as prop. No need to call `getSite`, id is available in `site.ID`.

Your `mapStateToProps` function needs to handle both cases. Something like this:
```js
function mapStateToProps( state, ownProps ) {
  const siteId = ownProps.siteId || ownProps.site.ID;
  const site = siteId ? getSite( state, siteId ) : ownProps.site;

  return {
    siteId,
    site,
    isPreviewable: isSitePreviewable( state, siteId ),
    siteSlug: getSiteSlug( state, siteId ),
  };
}
```

jsnajdr commented on client/blocks/site/index.jsx:
```diff
@@ -148,13 +178,32 @@ class Site extends React.Component {
+		isPreviewable: isSitePreviewable( state, siteId ),
+		siteId,
+		site,
+		siteSuffix: site ? '/' + site.slug : '',
```
We usually call this prop `siteSlug`. And for `<Site>`, a valid site is always available, so there doesn't need to be any default for null `site`. That's needed only in sidebar, which handles the "all sites selected" case.

jsnajdr commented on client/blocks/site/index.jsx:
```diff
@@ -68,6 +73,33 @@ class Site extends React.Component {
+			return;
+		}
+
+		event.preventDefault();
```
The `preventDefault` call is not needed. `button` doesn't have any default action to be prevented. This line is only useful for, e.g., links that have both onclick handler and a href attribute:
```jsx
<a href="/view/site.slug" onClick={ handleClick }>View</a>
```
Then the `handleClick` handler can prevent the `href` navigation from occuring.

jsnajdr commented on client/blocks/site/index.jsx:
```diff
@@ -68,6 +73,33 @@ class Site extends React.Component {
+			return;
+		}
+
+		if ( event.altKey || event.ctrlKey || event.metaKey || event.shiftKey ) {
```
The purpose of this code is to implement a neat feature for a `<a href>` link: if the link is clicked, JavaScript will open a preview modal with iframe. The `href` navigation is not performed. However, when the link is clicked with Ctrl-click, the intent is to open the link in new tab. Then the JavaScript doesn't intercept and the default browser behavior for `href` click is performed. If the `href` contains the site's frontend URL, it opens the site in new tab!

For `<button>`, this code is useless. But you can implement the behavior above by switching back to `<a>`. The linter will not yell at you any more, because the `a` element will have a valid `href` attribute.

Commit:
Updates based on @jsnajdr's excellent feedback

- Reverted to <a> tag from <button>
- Updated mapStateToProps()
- Removed dependencies that are no longer needed
- Updated siteSuffix to be siteSlug
- Removed CSS changes that are no longer needed

davemart-in commented:
@jsnajdr thank you for the thorough review! I had a hunch that a lot of that code could be done away with, but I wasn't sure, so I started off by just bringing everything over. Thanks for highlighting all of the areas where this could be improved and thanks for taking the time to share a code example!

Updates based on your feedback can be found in: https://github.com/Automattic/wp-calypso/commit/4ee13d750988955aa46a4d0c95c068fae81e2acc

Commit:
Add focus styles for keyboard accessibility

jsnajdr commented on client/blocks/site/style.scss:
```diff
@@ -71,8 +71,15 @@
-	&:focus {
+	.accessible-focus &:focus {
 		outline: none;
+		box-shadow: inset 0 0 0 2px var( --color-primary-light );
```
The focus ring is overlapped by the long-content-fade on the right side, with an unfavorable result:

<img width="282" alt="Screenshot 2019-04-26 at 15 00 41" src="https://user-images.githubusercontent.com/664258/56809456-2cc1d780-6834-11e9-8401-105fe599a6ce.png">

davemart-in replied:
Nice catch! Updated in: https://github.com/Automattic/wp-calypso/commit/dd1f099f8760ed332fc504be86b7c927f3ea41b9

Here's a quick preview of what it looks like now:

![](https://d1jfzjx68gj8xs.cloudfront.net/items/2A2z093W0e3p1j0d1n3v/Screen%20Recording%202019-04-26%20at%2010.10%20AM.gif)

jsnajdr commented on client/blocks/site/style.scss:
```diff
@@ -71,8 +71,15 @@
+			top: 2px;
+			right: 2px;
+			bottom: 2px;
+		}
```
What is the purpose of this `::after` element? The "Elements" devtool doesn't show it at all. These pseudo-element only become active when they have a `content` attribute, so this CSS in its current form seems to be a noop.

davemart-in replied:
Thanks! Removed in: https://github.com/Automattic/wp-calypso/commit/dd1f099f8760ed332fc504be86b7c927f3ea41b9

jsnajdr replied:
Thanks! Can you also add a comment explaining that this CSS is overriding the `long-content-fade`? It's hard to figure out otherwise.

Also, `display: none` should do the same job as `z-index: -1` and is more semantic.

davemart-in replied:
Can do! Updated in: https://github.com/Automattic/wp-calypso/commit/c244bf386d046c901ca27fb8b6a874466dbfedf4

jsnajdr reviewed (commented):
Thanks @davemart-in for making the suggested changes. The site preview now works very well. I posted a few more nits, much less serious than the first round.

Commit:
Removed trackMenuItemClick and brought everything inline based on @jsnajdr's recommendation

davemart-in commented on client/blocks/site/index.jsx:
```diff
@@ -68,6 +70,33 @@ class Site extends React.Component {
+		page( '/view/' + siteSlug );
+	};
+
+	trackMenuItemClick = menuItemName => {
```
Ohh... I love it. Updated in: https://github.com/Automattic/wp-calypso/commit/078d63791252820fb9e53fbf936fa37b6358e6b9

Commit:
Updated focus and ::After CSS based on @jsnajdr's feedback

Preview: https://cl.ly/10701ace3a85

shaunandrews commented:
Two things:

- When I click the view button in the sidebar, the preview loads but I'm unable to scroll. (Safari latest)
- Should the view button have an active state, like the other menu items do?

jsnajdr commented on client/blocks/site/index.jsx:
```diff
@@ -102,7 +125,7 @@ class Site extends React.Component {
 							: site.domain
 					}
-					onClick={ this.onSelect }
+					onClick={ this.onViewSiteClick }
```
The `Site` component is reused at a plenty of places and overriding the `onClick` handler breaks many of them. The `onSelect` prop should not be ignored. On `homeLink` should trigger the view behavior:
```js
onClick={ this.homeLink ? this.onViewSiteClick : this.onSelect }
```

Test by going to `/plans` and choosing a site from the site selector. Click should perform the select and navigate to the site's Plans page. At this moment, it navigates to `/view/site.blog` 👎

davemart-in replied:
Nice catch! Fixed in: https://github.com/Automattic/wp-calypso/commit/e3e2154f145f98f5277836f0394c31319f05b0f8

Commit:
Merge branch 'master' into update/inline-view-site

davemart-in commented:
@shaunandrews 

> When I click the view button in the sidebar, the preview loads but I'm unable to scroll. (Safari latest)

I DM'ed you in Slack about this. I can't even get Calypso to load for me locally on Safari. :-( 

> Should the view button have an active state, like the other menu items do?

I think so yeah, only `is-selected` is not being populated properly [here](https://github.com/Automattic/wp-calypso/blob/b0e35f6743b6cc0646cc1ec1d92b011b816d5c48/client/blocks/site/index.jsx#L109). If I can get a hand from @griffbrad or @jsnajdr to populate that value, I'm happy to add an active state in the CSS.

Commit:
Fixed onclick bug brought up by @jsnajdr

shaunandrews commented:
> I can't even get Calypso to load for me locally on Safari.

Not sure what the problem is there, but I've doubled checked in Chrome and Firefox and the scrolling issue I mentioned seems to be only in Safari. To make it more frustrating and mysterious, sometimes the scrolling will "fix" itself randomly. I can regularly reproduce the lack of scrolling, but I've no clue what causes Safari to "fix" itself — I can't see any classNames changing. I'm leaning towards a browser quirk.

As this likely its a regression caused by any of your changes, I think it'd be ok to move forward — perhaps we can create a new issue to track and (hopefully) fix the scrolling.

Commit:
Add comment and update CSS to hide ::after blur

jsnajdr commented:
> I think so yeah, only is-selected is not being populated properly here. If I can get a hand from @griffbrad or @jsnajdr to populate that value, I'm happy to add an active state in the CSS.

The `isSelected` and `isHighlighted` props are used in the `SiteSelector` list UI to show the selected site in dark gray and the mouseover-ed (i.e., highlighted) site in white:

![Screen Capture on 2019-05-01 at 18-26-02](https://user-images.githubusercontent.com/664258/57062157-a720aa80-6cbf-11e9-8508-37138f8f9710.gif)

If you want to show an active state on the view button, to make it fit better with the surrounding sidebar, reusing `isSelected` might not be a good idea -- both concepts are very different. A new prop, e.g., `isHomeLinkActive`, looks like a better option.

jsnajdr reviewed (approved):
I have no further objections :wink: Looks ready to :ship: Thanks @davemart-in for a fruitful collaboration.

davemart-in commented:
Cool. I spoke with @shaunandrews. I'm going to go ahead and deploy this as is.

Scrolling thing seems to be Safari specific. I've tested multiple times on Chrome and FireFox.

The other one is impossible at the moment because no CSS class is being added to that site card when you're on the site preview screen. Shaun offered to take a look at this at some point post deploy.

Commit:
Merge branch 'master' into update/inline-view-site

Commit:
Merge branch 'master' into update/inline-view-site
